### ISSUE TITLE
Support HTTP connection protocol upgrades thru Envoy

### ISSUE THREAD

**Issue by blake** [hashicorp/consul#8283]
#### Feature Description

HTTP supports a [protocol upgrade mechanism](https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism) which allows upgrading existing connections to a different protocol such as WebSocket. Envoy supports [WebSocket and HTTP upgrades](https://www.envoyproxy.io/docs/envoy/v1.14.3/intro/arch_overview/http/websocket), however this is not enabled by default. It requires providing an [upgrade_config](https://www.envoyproxy.io/docs/envoy/v1.14.3/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-upgrade-configs) to the HTTP Connection Manger (see https://github.com/hashicorp/consul/commit/fd993ccebb37ebdf5c08450da44a5ce8e22c6b30 for example).

This setting is not configurable in Consul as of 1.8.0, and because of this it is not possible to establish WebSocket connections through either ingress gateways or sidecar proxies when the service's configured [protocol](https://www.consul.io/docs/agent/config-entries/service-defaults.html#protocol) is `http`.

```shell
curl 'http://dashboard.ingress.consul/socket.io/?EIO=3&transport=websocket' \
  --include \
  --header 'Origin: http://dashboard.ingress.consul' \
  --header 'Connection: keep-alive, Upgrade' \
  --header 'Upgrade: websocket' \
  --header 'Sec-WebSocket-Version: 13' \
  --header 'Sec-WebSocket-Key: O1b2cWbnz4bk8LyYKDiE3Q=='

HTTP/1.1 403 Forbidden
date: Wed, 08 Jul 2020 06:29:45 GMT
server: envoy
connection: close
content-length: 0
```
It would be useful if Consul either enables WebSocket connection upgrades by default, or provides a way for operators to enable this via configuration.

#### Use Case(s)

Enables any application utilizing WebSockets to work through Ingress gateways or sidecar proxies.

Below is a successful connection upgrade request after deploying a custom version of Consul which configures Envoy with the appropriate upgrade config.

```shell
curl 'http://dashboard.ingress.consul/socket.io/?EIO=3&transport=websocket' \
  --include \
  --header 'Origin: http://dashboard.ingress.consul' \
  --header 'Connection: keep-alive, Upgrade' \
  --header 'Upgrade: websocket' \
  --header 'Sec-WebSocket-Version: 13' \
  --header 'Sec-WebSocket-Key: O1b2cWbnz4bk8LyYKDiE3Q=='

HTTP/1.1 101 Switching Protocols
upgrade: websocket
connection: Upgrade
sec-websocket-accept: f+WBHj+skiDNPZZlqh49NRVZo38=
date: Wed, 08 Jul 2020 07:52:15 GMT
server: envoy
content-length: 0
```

**Comment by darkedges** [comment]
@blake Could you share your changes so that I can get a working version for a demo I am wanting to do.

**Comment by blake** [comment]
Hi @darkedges,

I put a link to the commit SHA, fd993cc, in the description. The corresponding branch is https://github.com/hashicorp/consul/tree/blake/v1.8.0-ingress-websocket-fix.

You can check out this branch and build your own binary using the instructions in [hashicorp/consul/.github/CONTRIBUTING.md#building-consul](https://github.com/hashicorp/consul/blob/master/.github/CONTRIBUTING.md#building-consul). It _should_ work without issue, but I can't guarantee that. You may also want to rebase this on top of 1.8.2 since the latest release contains a few important fixes.

Hope this helps.

**Comment by timotheenicolas** [comment]
Hi really interested to see this fix merged in master

Thanks

**Comment by jason-man-blockone** [comment]
We'd like to see this feature in Consul too as we use websockets extensively in our application.

**Comment by kanak-kabara** [comment]
Would be nice to see this merged in.

**Comment by pblumo** [comment]
Will be great to merge in master this one . thanks

**Comment by awarner-greshamtech** [comment]
We are evaluating Nomad/Consul Connect/Vault and use websockets too, this is currently a blocker

**Comment by woz5999** [comment]
this is blocking me from using consul connect with some of our apps that require websockets. TCP service protocol isn't really an option because we're using hostnames to route.

**Comment by agrahamlincoln** [comment]
would love to see support for this, currently this makes our pilot use-case for consul service mesh a non-starter 😢

**Comment by Legogris** [comment]
FWIW submitted in https://github.com/hashicorp/consul/pull/9639

**Comment by dm-intropic** [comment]
@markan @Legogris any updates on if/when this would be added ?

**Comment by johnalotoski** [comment]
Would also like to see this supported.

**Comment by dekimsey** [comment]
tl;dr I'm bumping this issue as it's been awhile and the lack of support is painful. I'd like to see websocket support be implemented because there isn't any other good option for being able to configure it right now.

I ran into this issue with an http service we have. Since websocket support isn't currently enabled/exposed from Connect. I attempted to solve it using the `envoy_public_listener_json` escape hammer but found that to be unsatisfactory. It is a _really big_ hammer. Once you invoke it, the only thing Consul takes responsibility for is prefixing auth on the routes. All other parameters that Consul manages on your behalf like bind_address or port selection are now skipped or ignored. (I acutally think this is a gap and perhaps it could go set undefined values like these.)

For instance in my testing changing a running service using this parameter doesn't update Consul's notion of the service port in the sidecar registration. Since there is no templating behavior, there's no way to reasonably grab that information (or ideally have Consul set it if it is not defined). So if one is using this option, one is now responsible for manually assigning the service ports to the sidecars as Consul opt-outs. And these ports aren't assigned until after the registration occurs once on a node, a bit of a catch-22 there.

At the end I had to set the service as a tcp protocol and lose all L7 observability and routing behavior. 

I wanted to lay out my experience to demonstrate the current workarounds aren't great (or at least I had a hard time with them).

**Comment by porterctrlz** [comment]
Any news on this?

**Comment by georghildebrand** [comment]
Any update on this? It would be really nice to get WS working!

**Comment by kirooshu** [comment]
Same for us, we really need this feature as well so we can have WebSockets support with the Ingress Gateway.

**Comment by dani** [comment]
In my case, it's for Grafana and Loki, which both use a websocket for live logs view (with no fallback to polling or other mecanism). Exposing them as a TCP service instead of HTTP would hardly be a valid work around as I'm using the native Connect support in Traefik to expose those services, and this feature only supports HTTP.

**Comment by timotheencl** [comment]
Any news on this feature ?

**Comment by andersontechnetverrum** [comment]
Any update on this? It would be really nice to get WSS working!

**Comment by quinndiggitypolymath** [comment]
Just spent an entire day tracing down the source of a problem caused entirely by this issue with consul/envoy.

Throughout our codebases we have *hundreds* of these comments:

```
    #Protocol         = "http" # broken websocket connections - https://github.com/hashicorp/consul/issues/8283
    Protocol         = "tcp"
```

Enabling websocket connectivity, on an existing and otherwise working perfectly fine service, meant digging through a dozen layers of interconnected systems to track down the culprit that was breaking the websocket handshake.

Incredibly frustrating to finally remember `wait - was the layer 7 and all related functionality disabled for the consul/envoy websocket defect...` after an effectively wasted day of productivity.

**Comment by andersontechnetverrum** [comment]
End of the day this is a big show stopper to expand consul connect use cases. 

[quinndiggitypolymath](https://github.com/quinndiggitypolymath)  have you find a way to make web sockets working?

**Comment by quinndiggitypolymath** [comment]
@andersontechnetverrum, the tldr is:
```
resource "consul_config_entry" "service_defaults_ABC" {
  kind        = "service-defaults"
  name        = "ABC"
  config_json = jsonencode({
    Protocol = "tcp"
    ...
  })
}
```
ie. handle websocket services on layer 4 instead of 7 if you need to use consul/envoy - otherwise, just go with nginx, and avoid the envoy/consul side of things; it will save you a lot of time, in addition to just generally being more reliable, mature, and from a time when uber/lyft/facebook engineers hadn't yet made their contributions of pollution to the tech ecosystem.

**Comment by porterctrlz** [comment]
> End of the day this is a big show stopper to expand consul connect use cases.
> 
> [quinndiggitypolymath](https://github.com/quinndiggitypolymath) have you find a way to make web sockets working?

This stopped is in our tracks as well.

**Comment by andersontechnetverrum** [comment]
[quinndiggitypolymath](https://github.com/quinndiggitypolymath) replacing http protocol  to receive "tcp" worked! Thank you so much I was able to use WSS over consul connect integrated via NOMAD today. 
     
     I am using NOMAD integrate with Consul as backend here. Here is the template stanza working with 02 upstreams.  

    service {
      name = "${service}"
      canary_tags = ["urlprefix-${host}/canary strip=/canary"]
      tags = ["urlprefix-${host}/"]
      port = "3000" #whatever port number your application is allocating
      connect {
        sidecar_service = {
          tags = ["tcp"]
          proxy {
              config {
              protocol = "tcp"
            }
            upstreams {
              local_bind_port = "${upstreams_local_bind_port}"
              destination_name = "${environment}-${upstreams_destination_name_A}"
            }
            upstreams {
              local_bind_port = "${upstreams_local_bind_port_iot}"
              destination_name = "${environment}-${upstreams_destination_name_B}"
            }
          }
        }
                
        sidecar_task {
          name = "connect-proxy-${app}"

          driver = "docker"

          config {
            auth_soft_fail = true
            args = [
              "-c",
              "$${NOMAD_SECRETS_DIR}/envoy_bootstrap.json",
              "-l",
              "$${meta.connect.log_level}",
              "--concurrency",
              "$${meta.connect.proxy_concurrency}",
              "--disable-hot-restart"
            ]
          }

          logs {
            max_files     = 2
            max_file_size = 2 # MB
          }

          resources {
            cpu    = ${sidecar_task_resources_cpu} # MHz
            memory = ${sidecar_task_resources_memory} # MB
          }

          shutdown_delay = "5s"
        }

      }
    }

**Comment by Lord-Y** [comment]
@jkirschner-hashicorp is there any way you can look at this?
Seems like envoy need "only":
```
upgrade_configs:
- upgrade_type: websocket
```
From my enterprise, I will make a request to the support team.

**Comment by mr-miles** [comment]
I was pointed at this issue from the forums, when my websocket service mysteriously didn't work properly.

I want to expose a service to the outside world that provides some functionality over websockets, but also is part of a rest api via a service router.  After a lot of debugging I found that it required envoy configuration and that there wasn't an obvious way to add it via consul config.

My first question is - with the newer envoy patching mechanisms and escape-hatch config options, is there a way to add this into the generated config?  Ultimatly, the necessary addition to upgrade_configs as per Lord-Y's post above looks very simple.

Secondly, I can see there have been some PR submissions to solve this issue in the codebase.  What has prevented those being merged in?  I would be happy to dust off this work if it helps get it over the line.

Lastly, depending on the answer to that, could I ask that you update the documentation to cover that websocket upgrade requests will only succeed over tcp the mesh (if that were clear it would have saved me and others a lot of time)

Thanks for all your help on this

**Comment by simonctrlz** [comment]
I'm curious if this PR enables us to patch the Envoy config to enable protocol upgrades:
https://github.com/hashicorp/consul/pull/17487

**Comment by jkirschner-hashicorp** [comment]
Hi everyone,

Thank you for consistently expressing your interest in this issue and offering to contribute to address it.

Originally, we declined the approach proposed in PR #9639 because we weren't sure whether operators would be comfortable with websocket support being always enabled out-of-the-box (e.g., security implications). However, after much consideration, we have reversed that decision and recently merged the original approach proposed by @blake.

This change will be available in the upcoming patch releases:
- 1.14 release series: 1.14.9+
- 1.15 release series: 1.15.5+
- 1.16 release series: 1.16.1+
- Future major versions (1.17+)
